packages/core/src/Tool.ts

```diff
@@ -0,0 +1,139 @@
+/**
+ * A unique string identifier for each type of supported monorepo tool.
+ */
+export type ToolType = "bolt" | "lerna" | "pnpm" | "rush" | "singlePackage" | "yarn";
+
+/**
+ * A package.json access type.
+ */
+export type PackageAccessType = "public" | "restricted";
+
+/**
+ * An in-memory representation of a package.json file.
+ */
+export type PackageJSON = {
+    name: string;
+    version: string;
+    dependencies?: {
+        [key: string]: string;
+    };
+    peerDependencies?: {
+        [key: string]: string;
+    };
+    devDependencies?: {
+        [key: string]: string;
+    };
+    optionalDependencies?: {
+        [key: string]: string;
+    };
+    private?: boolean;
+    publishConfig?: {
+        access?: PackageAccessType;
+        directory?: string;
+        registry?: string;
+    };
+};
+
+/**
+ * An individual package json structure, along with the directory it lives in,
+ * relative to the root of the current monorepo.
+ */
+export type Package = {
+    /**
+     * The pre-loaded package json structure.
+     */
+    packageJson: PackageJSON;
+
+    /**
+     * Path to the directory containing this package, relative to the monorepo root (for
+     * a "root package", this is the string ".").
+     */
+    relativeDir: string;
```

Comment:
Last minute thought - `relativeDir` is useful for display purposes (for error messages and stuff like that) but for everything else, it's less useful as it's not sufficient for manipulating the package (you always have to `join` it with the `rootDir`). Perhaps it would be better to have an absolute path here? 🤔

Reply:
This is a tough call because no matter what choice you make, you are visiting pain upon the opposing side -- doing `path.relative()` to get a display path feel onerous, but so does doing `path.join()` to get an absolute path.

In Rush, the (equivalent) of this object solves this by providing both -- a `projectFolder` for the absolute path and the `relativeProjectFolder` for the relative path.

Maybe it makes sense to do the same here, bake in both paths in the returned object and then you don't have to choose?

Reply:
I've thought about this some more, and while I finish up remaining references in manypkg, I'm realizing that the two are not as equivalent as I claimed.

The difference is: if we go with absolute everything, you _can_ deduce the relative path; but the opposite is not true (given only a Package object).

This has pretty big implications for API surface area. For example, if you write a method that takes a single `Package` object, that method can read/update the json, write out updates, search for sibling packages, etc; none of that is possible if `Package` contains only a relative directory, because you don't actually know where it is.

I think, then, the right call is to go back to `dir` for all packages. (But we can still have `rootDir` at the Packages level.)

I'm a bit bummed about the `Packages` name when I see it used in other modules because it doesn't make a lot of sense. `PackageCollection` would be better, but _really_, it's a `Monorepo` -- a root dir, a collection of packages, an optional root package, a tool implementation, that's a monorepo. The fact that it _can_ be used to represent a single package maybe shouldn't dissuade us from just calling it that.